plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1033,6 +1027,99 @@ \section{Examples}
 % to \verb|ifix|.
 
+\section{Untyped Plutus Core}
+\label{sec:untyped-plc}
+For code transmission and on-chain execution it is important to have
+small programs.  In this section we define an untyped version of
+Plutus Core together with a type erasure function converting typed
+Plutus Core terms to untyped ones.
+
+As mentioned in Section~\ref{sec:blockchain-issues}, it is important to
+limit the consumption of computational resources in on-chain code.  A
+future version of this document will include a cost model which will
+provide fine-grained costs for individual operations and built-in
+functions, enabling accurate (dynamic) monitoring of execution costs.
+
+\subfile{figures/UntypedGrammar}
+
+\bigskip
+\noindent The grammar is essentially that of typed Plutus Core with all types
+and type-related operations discarded.  However, we require two new
+items: \texttt{delay} and \texttt{force}.  These are needed because
+type-level abstraction (\texttt{abs}) and application (\texttt{\{\}})
+have a computational effect.  The body of a type abstraction will not
+be executed until the type is instantiated; thus we cannot just
+replace $\abs{\alpha}{K}{M}$ with $M$, since the latter is evaluated
+immediately.  As we will see below, \texttt{delay} and \texttt{force}
+are used to preserve the effects of type abstraction and
+instantiation.
+
+As with typed Plutus Core, we can show that values (terms which can
+not be reduced) have a particular syntactic form: see
+Figure~\ref{fig:untyped-values}.
+
+%% The introduction of \texttt{delay} and \texttt{force} may be
+%% controversial since $\delay{M}$ can be implemented as $\lamU{u}{M}$
+%% and $\force{M}$ as $\appU{M}{\lamU{x}{x}}$.  However, doing this
+%% has a cost in code size.  Experiments show that
+%% adding \texttt{delay} and \texttt{force} increases the size of
+%% compressed validators by about 10\%, but using $\lamU{u}{M}$ and
+%% $\appU{M}{\lamU{x}{x}}$ increases sizes by about 20\%.  A third
+%% option is to introduce a global variable $u$ of type \texttt{unit}
+%% and replace $\force{M}$ with $\appU{M}{u}$: this only increases
+%% sizes by about 12\%, but complicates specification because programs
+%% have to be wrapped in an expression binding $u$ (and maybe the CEK
+%% machine has to start with an environment binding $u$ to
+%% $\lamU{x}{x}$).  Another option would be to extend the grammar with
+%% a unit term, but then it would be tempting to add terms
+%% for \texttt{true}, \texttt{false}, and so on.  It's possible that
+%% in actual validator code (especially after type
+%% normalisation), \texttt{delay} and \texttt{force} will typically
+%% appear in expressions of the form $\force{\delay{e}}$, which could
+%% immediately be replaced by $e$ alone, and then using $\lamU{x}{x}$
+%% might not cause so much overhead. I'll look into this.  Later:
+%% nope.  I looked at the on-chain validators for the use cases and
+%% this only happens about 10\% of the time.
```

Comment:
Perhaps.  I was leaving it there so I could whip it out if anyone complained about force/delay.

Reply:
I tried making some of it into a footnote, but that completely messed up the formatting of the rest of the document.  I've left the footnote there but commented in out in case later revision gives us space to restore it.